### PR TITLE
Buildmode Refactor

### PR DESCRIPTION
**Pull request by AnturK**
- Adds generic simple click intercept to clients
- Buildmode now suggest paths in advanced mode
- Copy-Paste Mode added. This is not guaranteed to create perfect copies.

### CODE REVIEW

File: code/modules/admin/verbs/buildmode.dm [hunk 1]
```diff
@@ -3,215 +3,269 @@
...
-/proc/build_click(mob/user, buildmode, params, obj/object)
-	var/obj/effect/bmode/buildholder/holder = null
-	for(var/obj/effect/bmode/buildholder/H)
-		if(H.cl == user.client)
-			holder = H
-			break
-	if(!holder) return
-	var/list/pa = params2list(params)
+/datum/buildmode/proc/ClickOn(mob/user,params,A)
+	world << "[user] clicking on [A]. Interceptin"
```
Comment (RemieRichards) [inline code/modules/admin/verbs/buildmode.dm]: debug

File: code/_onclick/ai.dm [hunk 2]
```diff
@@ -27,9 +27,9 @@
 		return
 	next_click = world.time + 1
 
-	if(client.buildmode) // comes after object.Click to allow buildmode gui objects to be clicked
-		build_click(src, client.buildmode, params, A)
-		return
+	if(client.click_intercept)
+		if(call(client.click_intercept,"ClickOn")(src,params,A))
```
Comment (neersighted) [inline code/_onclick/ai.dm]: Spaces in proc calls please

File: code/_onclick/ai.dm [hunk 3]
```diff
@@ -10,9 +10,9 @@
 	Note that AI have no need for the adjacency proc, and so this proc is a lot cleaner.
 */
 /mob/living/silicon/ai/DblClickOn(var/atom/A, params)
-	if(client.buildmode) // comes after object.Click to allow buildmode gui objects to be clicked
-		build_click(src, client.buildmode, params, A)
-		return
+	if(client.click_intercept)
+		if(call(client.click_intercept,"ClickOn")(src,params,A))
```
Comment (neersighted) [inline code/_onclick/ai.dm]: Ditto

File: code/modules/admin/verbs/buildmode.dm [hunk 4]
```diff
@@ -3,215 +3,269 @@
...
+	var/right_click = pa.Find("right")
+	var/left_click = pa.Find("left")
+	var/alt_click = pa.Find("alt")
+	var/ctrl_click = pa.Find("ctrl")
+
+	. = 1
+	switch(mode)
 		if(BASIC_BUILDMODE)
-			if(istype(object,/turf) && pa.Find("left") && !pa.Find("alt") && !pa.Find("ctrl") )
+			if(istype(object,/turf) && left_click && !alt_click && !ctrl_click )
```
Comment (neersighted) [inline code/modules/admin/verbs/buildmode.dm]: No need for a space here

File: code/modules/client/client defines.dm [hunk 5]
```diff
@@ -4,7 +4,7 @@
 		//ADMIN THINGS//
 		////////////////
 	var/datum/admins/holder = null
-	var/buildmode		= 0
+	var/click_intercept = null // Needs to implement ClickOn(user,params,atom) proc
```
Comment (neersighted) [inline code/modules/client/client defines.dm]: `var/datum`?

File: code/modules/admin/verbs/buildmode.dm (hunk 1 above)
Comment (AnturK) [inline code/modules/admin/verbs/buildmode.dm]: Whoops

**Comment by Shadowlight213** [comment]
:+1: :x: :100: to the path suggestion in advanced mode. That was probably one of the major annoyances with it as you had to type the path perfectly

**Comment by lzimann** [comment]
pretty good and useful, thanks :+1:

File: code/modules/admin/verbs/buildmode.dm [hunk 6]
```diff
@@ -3,215 +3,265 @@
...
+			user << "\blue Right Mouse Button on buildmode button = Select generator"
+			user << "\blue ***********************************************************"
+		if(COPY_BUILDMODE)
+			user << "\blue ***********************************************************"
+			user << "\blue Left Mouse Button on obj/turf/mob   = Spawn a Copy of selected target"
+			user << "\blue Right Mouse Button on obj/mob = Select target to copy"
+			user << "\blue ***********************************************************"
+
+
+/datum/buildmode/proc/pick_closest_path(value)
```
Comment (tkdrg) [inline code/modules/admin/verbs/buildmode.dm]: Doesn't spawn do this as well? It could probably be moved up to a helper.

Comment (AnturK) [inline code/modules/admin/verbs/buildmode.dm]: Yep, will do that.

**Comment by StormAGeddonz** [comment]
:thumbsup: 👍👍👍👍👍👍

**Pull request merged**